File: NormalDataFrame.py
# Regular DF -----------------------------------------------------------------------------------------------------------
import pandas as pd
from timeit import default_timer as timer

import util


def build(messages):
    temp = timer()
    texts = []
    dates = []
    times = []
    authors = []
    for index in range(len(messages)):
        # FORMAT: "mm/dd/yy, HH:MM - [Author]: [message]"
        message = messages[index]
        date = message[:message.find(',')]
        time = message[message.find(', ') + 2:message.find(' - ')]
        author = util.remove_unwanted_chars(message[message.find(' - ') + 3:util.find_2nd(message, ':')])
        text = util.remove_unwanted_chars(message[util.find_2nd(message, ':') + 1:])
        if util.find_2nd(message, ':') == -1:
            author = "WhatsApp"
            text = util.remove_unwanted_chars(message[message.find(' - ') + 3:])
        elif "created group" in author or "changed the" in author:
            text = author
            author = "WhatsApp"
        texts.append(text)
        times.append(time)
        dates.append(date)
        authors.append(author)
        print("NDF -", index, end='\r')

    util.append_time(timer() - temp)
    return pd.DataFrame(list(zip(dates, times, authors, texts)),
                        columns=['date', 'time', 'author', 'message'])


def row(df, i):
    return (df.at[i, "date"],
            df.at[i, "time"],
            df.at[i, "author"],
            df.at[i, "message"])


def get_authors(df):
    authors = []
    for i in range(len(df)):
        date, time, author, message = row(df, i)
        if author not in authors:
            authors.append(author)
    return authors


def avg(nd):
    try:
        return sum([len(string) for string in nd['message']]) / len(nd)
    except ZeroDivisionError:
        return 0


def count(nd):
    return len(nd)

File: NormalDataFrame.java
public class NormalDataFrame extends DataFrame {
    public NormalDataFrame(String[] a, String[] b, String[] c, String[] d, String columnA, String columnB, String columnC, String columnD) {
        super(new String[][]{a, b, c, d}, new String[]{columnA, columnB, columnC, columnD});
    }

    public NormalDataFrame(String[][] data, String[] columns) {
        super(data, columns);
    }

    public NormalDataFrame(NormalDataFrame nndf) {
        super(nndf);
    }

    public static NormalDataFrame build(String[] messages) {
        String[] texts = new String[messages.length];
        String[] dates = new String[messages.length];
        String[] times = new String[messages.length];
        String[] authors = new String[messages.length];
        for (int index = 0; index < messages.length; index++) {
            // FORMAT: "mm/dd/yy, HH:MM - [Author]: [message]"
            String message = messages[index];
            int comma = message.indexOf(", ");
            String date = message.substring(0, comma);
            String time = message.substring(comma + 2, message.indexOf(" - "));
            int secondColon = Util.find2nd(message, ":");
            String author;
            String text;
            if (secondColon > -1) {
                author = Util.removeUnwantedChars(message.substring(message.indexOf(" - ") + 3, secondColon));
                text = Util.removeUnwantedChars(message.substring(secondColon + 1));
            } else {
                author = "WhatsApp";
                text = message.substring(message.indexOf(" - ") + 3);
            }
            if (author.contains("created group") || author.contains("changed the")) {
                text = author;
                author = "WhatsApp";
            }
            texts[index] = text;
            times[index] = time;
            dates[index] = date;
            authors[index] = author;
        }
        return new NormalDataFrame(dates, times, authors, texts, "date", "time", "author", "message");
    }
}
